packages/xforms-engine/test/parse/xpath/dependency-analysis.test.ts

```diff
@@ -0,0 +1,337 @@
+
+			contextNodeset: '/data/grp',
+			expression: 'current()/foo/bar and current()/../quux',
+			expected: ['/data/grp/foo/bar', '/data/quux'],
+		},
+
+		{
+			description: 'No context; instance() and relative path; extracted and returned and as-is',
+
+			contextNodeset: null,
+			expression: 'instance("foo")/bar | ./quux',
+			expected: ['instance("foo")/bar', './quux'],
+		},
+
+		{
+			description:
+				'Relative context; instance() and relative path; extracted, instance call returned as-is, relative paths resolved to context',
+
+			contextNodeset: 'sel1',
+			expression: 'instance("foo")/bar and (bat or ./quux)',
+			expected: ['instance("foo")/bar', 'sel1/bat', 'sel1/quux'],
+		},
+
+		{
+			description:
+				'Relative context; instance() and relative path; extracted, instance call returned as-is, relative paths resolved to context',
+
+			contextNodeset: 'sel1',
+			expression: 'instance("foo")/bar and (bat or ../quux)',
+			expected: ['instance("foo")/bar', 'sel1/bat', 'sel1/../quux'],
+		},
+
+		{
+			description:
+				'Absolute context; instance() and relative path; extracted, instance call returned as-is, relative paths resolved to context',
+
+			contextNodeset: '/data/sel1',
+			expression: 'instance("foo")/bar and (bat or ../quux)',
+			expected: ['instance("foo")/bar', '/data/sel1/bat', '/data/quux'],
+		},
+
+		{
+			description:
+				'Context is instance("id") descendant. Expression references another instance (resolved as-is); relative sibling and relative child (resolved from context); relative sibling and child with current() calls (likewise resolved from context); absolute reference (resolved as-is)',
+
+			contextNodeset: 'instance("foo")/bar',
+			expression:
+				'if(position() > 2, instance("quux")/zig, ../zag | ./zag | current()/zag | current()/../zag) or /beep//boop',
+			expected: [
+				'instance("quux")/zig',
+				'instance("foo")/zag',
+				'instance("foo")/bar/zag',
+				'/beep//boop',
+			],
+		},
+
+		{
+			description:
+				'Context is absolute. Expression includes multiple references to instance($some-id); each is extracted and returned as-is',
+
+			contextNodeset: '/data/quux',
+			expression: 'instance("first")/foo or instance("second")/bar',
+			expected: ['instance("first")/foo', 'instance("second")/bar'],
+		},
+
+		{
+			description:
+				'No context. Expression with instance("id") reference has its own parent steps resolved',
+
+			contextNodeset: '/data/quux',
+			expression: 'instance("first")/foo/bar/../quux',
+			expected: ['instance("first")/foo/quux'],
+		},
+
+		{
+			description: 'Resolution without context: parent steps',
+
+			contextNodeset: null,
+			expression: '/data/foo/../bar and foo/bar/../bat or /data/quux/../zig',
+			expected: ['/data/bar', 'foo/bat', '/data/zig'],
+		},
+
+		{
+			description:
+				'Arbitrary FilterExpr with either a trailing Step or a Predicate (without any specialized handling for the FunctionName in use)',
```

Comment:
This is exercising https://github.com/getodk/web-forms/pull/166/files#diff-b4d9c386b1f7bee4beb508d52cf07bfb56e90fd4b2bc17845163cfa6075af4f4R168 I believe. And so if a function were known and not one of the special ones any paths used as arguments would be resolved and returned? Maybe that's a test to add.

Reply:
It's exercising that, yes.

> And so if a function were known and not one of the special ones any paths used as arguments would be resolved and returned? Maybe that's a test to add.

If you're asking for a test for specifically resolving the paths as Arguments of a FunctionCall—something like `any-non-nodeset-fn(foo, /bar, //etc/., ..)`—does [this case calling `if(...)`](https://github.com/getodk/web-forms/pull/166/files#diff-cb3208eef759b81b33d971b0f6ccea7c8f2c0c041fdc98b2be1d9b515855cc13R126) suffice? If not, can you be more specific about what kind of expression you don't see covered?

Reply:
Sweet! I just missed it, probably because my brain doesn't tend to accept that `if` is a function.

Reply:
Yeah. It might be a little easier to reconcile if you've worked in a lisp or something else where `if` can be implemented in userland (but even in a lisp it'll be a macro, not a function).